python/test_model/types.py

```diff
@@ -1129,28 +1135,156 @@ def __repr__(self) -> str:
+            and yardl.structural_equal(self.aliased_dynamic_nd, other.aliased_dynamic_nd)
+        )
+
+    def __str__(self) -> str:
+        return f"RecordWithGenericArrays(nd={self.nd}, fixedNd={self.fixed_nd}, dynamicNd={self.dynamic_nd}, aliasedNd={self.aliased_nd}, aliasedFixedNd={self.aliased_fixed_nd}, aliasedDynamicNd={self.aliased_dynamic_nd})"
+
+    def __repr__(self) -> str:
+        return f"RecordWithGenericArrays(nd={repr(self.nd)}, fixedNd={repr(self.fixed_nd)}, dynamicNd={repr(self.dynamic_nd)}, aliasedNd={repr(self.aliased_nd)}, aliasedFixedNd={repr(self.aliased_fixed_nd)}, aliasedDynamicNd={repr(self.aliased_dynamic_nd)})"
+
+
+class RecordWithGenericMaps(typing.Generic[T, U]):
+    m: dict[T, U]
+    am: AliasedMap[T, U]
+
+    def __init__(self, *,
+        m: typing.Optional[dict[T, U]] = None,
+        am: typing.Optional[AliasedMap[T, U]] = None,
+    ):
+        self.m = m if m is not None else {}
+        self.am = am if am is not None else {}
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, RecordWithGenericMaps)
+            and yardl.structural_equal(self.m, other.m)
+            and yardl.structural_equal(self.am, other.am)
+        )
+
+    def __str__(self) -> str:
+        return f"RecordWithGenericMaps(m={self.m}, am={self.am})"
+
+    def __repr__(self) -> str:
+        return f"RecordWithGenericMaps(m={repr(self.m)}, am={repr(self.am)})"
+
+
+class RecordContainingGenericRecords(typing.Generic[A, B, B_NP]):
     g1: RecordWithOptionalGenericField[A]
     g1a: RecordWithAliasedOptionalGenericField[A]
     g2: RecordWithOptionalGenericUnionField[A, B]
     g2a: RecordWithAliasedOptionalGenericUnionField[A, B]
     g3: MyTuple[A, B]
     g3a: AliasedTuple[A, B]
+    g4: RecordWithGenericVectors[B]
+    g5: RecordWithGenericFixedVectors[B]
+    g6: RecordWithGenericArrays[B_NP]
+    g7: RecordWithGenericMaps[A, B]
 
     def __init__(self, *,
-        g1: RecordWithOptionalGenericField[A],
-        g1a: RecordWithAliasedOptionalGenericField[A],
-        g2: RecordWithOptionalGenericUnionField[A, B],
-        g2a: RecordWithAliasedOptionalGenericUnionField[A, B],
+        g1: typing.Optional[RecordWithOptionalGenericField[A]] = None,
+        g1a: typing.Optional[RecordWithAliasedOptionalGenericField[A]] = None,
+        g2: typing.Optional[RecordWithOptionalGenericUnionField[A, B]] = None,
+        g2a: typing.Optional[RecordWithAliasedOptionalGenericUnionField[A, B]] = None,
         g3: MyTuple[A, B],
         g3a: AliasedTuple[A, B],
+        g4: typing.Optional[RecordWithGenericVectors[B]] = None,
+        g5: RecordWithGenericFixedVectors[B],
+        g6: RecordWithGenericArrays[B_NP],
+        g7: typing.Optional[RecordWithGenericMaps[A, B]] = None,
     ):
-        self.g1 = g1
-        self.g1a = g1a
-        self.g2 = g2
-        self.g2a = g2a
+        self.g1 = g1 if g1 is not None else RecordWithOptionalGenericField(v=None)
```

Comment:
Looks like we should be able to remove `v=None` here (and `v=[], av=[]` a few lines down)

Reply:
Trying to remove these superfluous args is making my brain hurt.

Right now, we pass defaults for every field in a generic records (closed or open) and we don't pass any defaults for fields in non-generic records.

See  https://github.com/microsoft/yardl/blob/41a9d22437c158253c7d8c1faeb75f63f923a8be/tooling/internal/python/types/types.go#L906C1-L921C6

When building the `args` slice, I don't see how to differentiate between fields that *need* to be instantiated vs. fields that don't need to be instantiated because they have a default value.

Reply:
Ha ha, this is what the code that removed in #96 was meant to handle (and I couldn't remember why):

```
		genericDef := st[t.GetQualifiedName()].(*dsl.RecordDefinition)

		args := make([]string, 0)

		for i, f := range t.Fields {
			fieldDefaultExpr, fieldDefaultKind := typeDefault(f.Type, contextNamespace, "", st)
			if fieldDefaultKind == defaultValueKindNone {
				return "", defaultValueKindNone
			}

			_, genDefaultKind := typeDefault(genericDef.Fields[i].Type, contextNamespace, "", st)
			if genDefaultKind == defaultValueKindNone {
				args = append(args, fmt.Sprintf("%s=%s", common.FieldIdentifierName(f.Name), fieldDefaultExpr))
			}
		}
```

Reply:
Hah! Now it makes perfect sense. Glad I got your eyes on it.